### PR TITLE
Implement transposition table with 2DEEP replacement scheme

### PR DESCRIPTION
**Pull request by Zhunaidm**

### CODE REVIEW

File: src/main/java/za/ac/sun/cs/adversarial/transposition/TranspositionEntry.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    private Move move;
+    private int score;
+    private int flag;
+    private int depth;
+
+    public TranspositionEntry(long key, Move move, int score, int flag, int depth) {
+        this.key = key;
+        this.move = move;
+        this.score = score;
+        this.flag = flag;
```
Comment (Zhunaidm) [inline src/main/java/za/ac/sun/cs/adversarial/transposition/TranspositionEntry.java]: We can make this an enum.

File: src/main/java/za/ac/sun/cs/adversarial/transposition/TranspositionEntry.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    public int getFlag() {
+        return this.flag;
+    }
+
+    public int getDepth() {
+        return this.depth;
+    }
+
+    public void setKey(int key) {
```
Comment (Zhunaidm) [inline src/main/java/za/ac/sun/cs/adversarial/transposition/TranspositionEntry.java]: Should be a `long` :+1:

File: src/main/java/za/ac/sun/cs/adversarial/transposition/TranspositionTable.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+public class TranspositionTable {
+    
+    private int bits; // Number of bits to store.
+    private int size; // Size of transposition table.
+    private TranspositionWrapper[] table;
+
+    public TranspositionTable(int bits) {
+        this.bits = bits;
+        this.size = 2 ^ bits;
```
Comment (Zhunaidm) [inline src/main/java/za/ac/sun/cs/adversarial/transposition/TranspositionTable.java:13]: Forcing some collisions :dagger:

File: src/test/java/za/ac/sun/cs/search/adversarialagent/TestTranspositionTable.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+import za.ac.sun.cs.adversarial.domain.Board;
+import za.ac.sun.cs.adversarial.domain.Move;
+import za.ac.sun.cs.adversarial.hash.Zobrist;
+import za.ac.sun.cs.adversarial.transposition.TranspositionTable;
+import za.ac.sun.cs.adversarial.transposition.TranspositionEntry;
+
+public class TestTranspositionTable {
```
Comment (Zhunaidm) [inline src/test/java/za/ac/sun/cs/search/adversarialagent/TestTranspositionTable.java]: Will add some tests to force / test collisions :zap:

**Review by Zhunaidm — COMMENTED** [review]
:+1:

**Comment by PhillipVH** [comment]
Looks good. Going onto `master` now! :zap:

**Pull request merged**